src/main/java/ui/components/pickers/LabelPickerState.java

```diff
@@ -0,0 +1,260 @@
+    private OptionalInt currentSuggestionIndex;
+
+    public LabelPickerState(Set<String> initialLabels, List<TurboLabel> allLabels, String userInput) {
+        this(initialLabels, new ArrayList<>(), new ArrayList<>(), new ArrayList<>(), allLabels,
+                OptionalInt.empty());
+        update(userInput);
+    }
+
+    private LabelPickerState(Set<String> initialLabels, List<String> addedLabels, List<String> removedLabels,
+                             List<String> matchedLabels, List<TurboLabel> allLabels, 
+                             OptionalInt currentSuggestionIndex) {
+        this.initialLabels = initialLabels;
+        this.addedLabels = addedLabels;
+        this.removedLabels = removedLabels;
+        this.matchedLabels = matchedLabels;
+        this.allLabels = allLabels;
+        this.currentSuggestionIndex = currentSuggestionIndex;
+    }
+
+    /**
+     * Updates current state based on given user input
+     * @param userInput 
+     */
+    private final void update(String userInput) {
+        List<String> confirmedKeywords = getConfirmedKeywords(userInput);
+        for (String confirmedKeyword : confirmedKeywords) {
+            updateIfMatchesLabel(confirmedKeyword);
+        }
+
+        Optional<String> keywordInProgess = getKeywordInProgress(userInput);
+        if (keywordInProgess.isPresent()) {
+            updateMatchedLabels(keywordInProgess.get());
+        }
+    }
+
+    /**
+     * Updates current state if there is at least one matching label based on the 
+     * given keyword
+     * @param keyword
+     */
+    private final void updateIfMatchesLabel(String keyword) {
+        if (TurboLabel.hasMatchedLabel(allLabels, keyword)) {
+            updateAssignedLabels(TurboLabel.getFirstMatchingTurboLabel(allLabels, keyword));
+        }
+    }
+
+    /**
+     * Updates assignedLabels based on properties of a label 
+     * @param label
+     */
+    public final void updateAssignedLabels(TurboLabel label) {
+        String labelName = label.getFullName();
+
+        if (isAnInitialLabel(labelName)) {
+            if (isARemovedLabel(labelName)) {
+                removeConflictingLabels(label);
+                removedLabels.remove(labelName);
+            } else {
+                removedLabels.add(labelName);
+            }
+        } else {
+            if (isAnAddedLabel(labelName)) {
+                addedLabels.remove(labelName);
+            } else {
+                // add new label
+                removeConflictingLabels(label);
+                addedLabels.add(labelName);
+            }
+        }
+    }
+
+    /**
+     * Updates the list of labels which labels' names contain the current keyword.
+     * This list of labels can then be retrieved later via getMatchedLabels()
+     *
+     * The suggestion index will be pointed to the first label that matches the keyword, if there is.
+     *
+     * @param keyword
+     */
+    private final void updateMatchedLabels(String keyword) {
```

Comment:
It felt like two things, although they are always done together: 1. update matched labels 2. update suggested label